akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala

```diff
@@ -0,0 +1,163 @@
+    val requestUriInterceptor = ProxyUriRequestInterceptor(uri).andThen(ProxyHeaderRequestInterceptor)
+    mapRequest(requestUriInterceptor).tflatMap { _ ⇒
+      forwardRequest(configOption)
+    }
+  }
+
+  def forwardRequest(configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
+    extractExecutionContext.flatMap { implicit ec ⇒
+      extractRequest.flatMap { request ⇒
+        getOrCreateDefaultConfig(configOption).flatMap { reverseProxyConfig ⇒
+          val responseFuture: Future[HttpResponse] = reverseProxyConfig.requestExecutor.apply(request)
+          reverseProxyConfig.circuitBreakerOption match {
+            case Some(circuitBreaker) ⇒ onCompleteWithBreaker(circuitBreaker)(responseFuture)
+            case None                 ⇒ onComplete(responseFuture)
+          }
+        }
+      }
+    }
+  }
+
+  private[akka] def getOrCreateDefaultConfig(configOption: Option[ReverseProxyConfig]): Directive1[ReverseProxyConfig] = {
+    configOption match {
+      case Some(config) ⇒ provide(config)
+      case None ⇒ {
+        extractMaterializer.flatMap { mat ⇒
+          extractActorSystem.flatMap { system ⇒
+            provide(ReverseProxyConfig(None, CachedHostConnectionPoolRequestExecutor()(system, mat)))
+          }
+        }
+      }
+    }
+  }
+}
+
+private[akka] case class ProxyUriRequestInterceptor(val uri: Uri) extends ((HttpRequest) ⇒ HttpRequest) {
+  override def apply(httpRequest: HttpRequest): HttpRequest = {
+    httpRequest.withUri(uri)
+  }
+}
+
+private[akka] case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
+  override def apply(httpRequest: HttpRequest): HttpRequest = {
+    val headers: Seq[HttpHeader] = httpRequest.headers
+    val remoteAddressHeaderOption = headers.collectFirst({ case h: `Remote-Address` ⇒ h })
+
+    val xRealIpPF: PartialFunction[HttpHeader, `X-Real-Ip`] = { case h: `X-Real-Ip` ⇒ h }
+    val xForwardedForPF: PartialFunction[HttpHeader, `X-Forwarded-For`] = { case h: `X-Forwarded-For` ⇒ h }
+
+    val xRealIpHeaderOption = headers.collectFirst(xRealIpPF)
+    val xForwardedForHeaderOption = headers.collectFirst(xForwardedForPF)
+    //TODO: Add `Via` HttpHeader as defined in https://tools.ietf.org/html/rfc2616#section-14.45 in akka.http.scaladsl.model.headers
+    //var viaHeaderOption: Option[HttpHeader] = headers.collectFirst({case h: HttpHeader if "via".equalsIgnoreCase(h.name) => h})
+
+    val remoteAddressOption = remoteAddressHeaderOption.map(_.address)
+    val xRealIpOption = xRealIpHeaderOption.map(_.address)
+
+    val ipOption = xRealIpOption.orElse(xForwardedForHeaderOption.flatMap(_.addresses.headOption).orElse(remoteAddressOption))
+
+    val updatedXRealIpHeaderOption = ipOption.map(`X-Real-Ip`(_))
+    val updatedXForwardedForHeaderOption = remoteAddressOption.map(remoteAddresss ⇒ xForwardedForHeaderOption.fold(`X-Forwarded-For`(remoteAddresss))(h ⇒ `X-Forwarded-For`(h.addresses ++ Seq(remoteAddresss))))
+
```

Comment:
you need to filter Connection and `Timeout-Access` headers (might be worth having a look over https://www.mnot.net/blog/2011/07/11/what_proxies_must_do).